flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,211 @@
+that are entirely local to the function, or are function parameters, may be modified as normal. So, for example, this code would be allowed:
+
+    ```cadence
+    view fun foo(): Int {
+        let a: [Int] = []
+        a.append(3)
+        return a.length
+    }
+    ```
+
+    while this would not:
+
+    ```cadence
+    let a: [Int] = []
+    view fun foo(): Int {
+        a.append(3)
+        return a.length
+    }
+    ```
+
+    Note that it is not always possible to determine statically whether or not a mutation is occuring on a locally or externally scoped variable, 
+as in this case, for example:
+
+    ```cadence
+    let a: [Int] = []
+    view fun foo(i: Int): Int {
+        let b: [Int] = []
+        let c = [a, b]
+        c[i].append(4) // it is impossible to tell statically whether it is `a` or `b` that is receiving the write here
+        return a.length
+    }
+    ```
+
+    Cases like this will have to be rejected statically for safety. Field reads, of course, will remain valid, as they do not modify any state. 
+
+* Writing to storage. This includes operations like `save` and `load` (since if a resource is loaded it will be moved out of storage), as well as `unlink`, but
+not operations like `copy`, `borrow`, `type`, or `getCapability`.
+
+* Writing to account state. This includes operations like adding, updating or removing a contract and adding, updating or removing account keys.
+
+* Logging data (calling the `log` function)
+
+* Emitting events
+
+* Mutating a reference. Any writes to references will have to be rejected, as it is generally not possible to know what value is being referenced. Consider code like the following:
+
+    ```cadence
+    view fun foo(a: [Int], b: bool): Int {
+        let c: [Int] = &[] as &[Int]
+        if b {
+            c = &a as &[Int]
+        }
+        c.append(4) // it is impossible to tell statically whether `a` is receiving the write here
+        return a.length
+    }
+    ```
+
+Any functions declared to be `view` will enforce that none of these operations occur in their bodies. If such an operation does occur, then the checker will report
+an error identifying the impure operation. Impure function declarations have no additional typechecking behavior; any operations are permitted.
+
+Note that the current design allows both emitting events and logging data from within a view function. These are traditionally considered side effects, but 
+it is perhaps worth some discussion on whether or not these should be allowed in `view` functions.  
+
+Purity also interacts covariantly with function subtyping: `view` functions are a subtype of impure functions. So, the following declarations would typecheck:
+
+```cadence
+    let a: (view (): Void) = view fun() {}
+    let b: ((): Void) = view fun() {}
+    let c: ((): Void) = fun() {}
+    let d: (((view (): Void)): Void) = fun foo(x:((): Void)) {} // contravariance
```

Comment:
It looks like the extra parentheses for the first parameter of the function type are not necessary.
Maybe also make the comment more descriptive:

```suggestion
    let d: ((view (): Void): Void) = fun foo(f: ((): Void)) {} // function parameters are contravariant
```